### PR TITLE
Combine water level data

### PR DESCRIPTION
**Pull request by limnoliver**
Combining the observed reservoir water level data from NWIS pulls, NYC DEP-provided data, and historical USGS data/reports provided by the NY WSC. Though Hayley has shown these sources are near identical, I kept data in this order: NWIS, NYC-DEP, then historical USGS. I chose the historical data last so that daily observed data would be prioritized over the monthly interpolated data from the historical USGS data. 

A few questions:

1) I think the next step is to interpolate between observations that come from the NYC DEP data in the early part of the record (see graph below). Unlike the monthly to daily interpolation where we just assumed the same value over the month, we would need to consider how to interpolate between irregularly sampled dates. Just linear interpolation between points? 
2) From the reservoir meeting notes, I think we decided to keep the estimated (from surface area) water level data out of the release?

```
dat <- readRDS('out_data/reservoir_level_obs.rds')

ggplot(dat, aes(x = date, y = surface_elevation_m)) +
    geom_point(aes(color = source, shape = data_type), alpha = 0.1) +
    facet_wrap(~site_id, nrow = 2, scales = 'free') +
    theme_bw()
```
image

### CODE REVIEW

**Comment by aappling-usgs** [comment]
Those historical data look a lot better than the remote-sensing-based surface elevations. I agree with leaving them out of the data release for now.

However, when we get to modeling reservoirs for which we have less data, we may need to lean on the remote-sensing estimates. For that effort we may want to run a comparison of Pepacton or Cannonsville models built on in-situ vs RS-based data to understand what we're getting into with the new reservoirs.

**Comment by limnoliver** [comment]
@aappling-usgs - you need to get to daily info, right? Should we linearly interpolate the irregularly sampled NYC data to get down to daily?

**Comment by aappling-usgs** [comment]
yes, please

**Comment by limnoliver** [comment]
Thinking that we would prefer to have interpolated daily values (from NYC DEP) versus the interpolated monthly average value (from historical USGS data) when we have the choice (see ~2008).

image

**Comment by limnoliver** [comment]
Ok - I filtered out average monthly values and interpolated the dailies instead. Here is the final product:

image

File: 2_observations.yml [hunk 1]
```diff
@@ -62,9 +63,32 @@ targets:
...
+      in_ind = '../lake-temperature-model-prep/7a_nwis_munge/out/NYC_DEP_reservoir_waterlevels.rds.ind',
+      in_repo = I('../lake-temperature-model-prep/'),
+      site_ids = reservoir_modeling_site_ids)
+      
+  out_data/reservoir_level_usgs_historical.rds:
+    command: fetch_filter_historical(
+      out_rds = target_name,
+      in_ind = '../delaware-model-prep/2_observations/out/interpolated_daily_reservoir_water_budget_components.csv.ind',
+      in_repo = I('../delaware-model-prep/'),
+      xwalk = I(c('Cannonsville' = 'nhdhr_120022743', 'Pepacton' = 'nhdhr_151957878')),
```
Comment (hcorson-dosch-usgs) [inline 2_observations.yml]: Just out of curiosity, why did you need to pass this xwalk argument in addition to the named `reservoir_modeling_site_ids` vector?

File: src/file_functions.R [hunk 2]
```diff
@@ -121,3 +121,45 @@ sf_to_zip <- function(zip_filename, sf_object, layer_name){
...
+  nyc <- readRDS(nyc_levels)
+  hist <- readRDS(hist_levels) %>% 
+    mutate(date = as.Date(date)) %>%
+    rename(surface_elevation_m = res_level_m) %>%
+    # filtering out monthly interpolated values since we now of NYC-DEP data that covers same period
+    filter(!data_type %in% 'monthly')
+  
+  obs_dat <- bind_rows(nwis, nyc, hist, .id = 'source') %>%
+    group_by(site_id, date) %>%
+    # sources were ordered by priority, so can filter on min source
```
Comment (hcorson-dosch-usgs) [inline src/file_functions.R:137]: This ordering happens when you provide arguments to `bind_rows()` in a specific order?

Comment (hcorson-dosch-usgs) [reply src/file_functions.R:137]: Aha - just ran the code line by line and now see that `.id` returns as an integer, which allows for `slice_min`. I was thinking `.id` returned nwis/nyc/hist

File: src/file_functions.R [hunk 3]
```diff
@@ -121,3 +121,45 @@ sf_to_zip <- function(zip_filename, sf_object, layer_name){
...
+    slice_min(source) %>%
```
Comment (hcorson-dosch-usgs) [inline src/file_functions.R:138]: Haven't seen `slice_min()` before -- cool how it pairs with the `group_by` call

File: src/file_functions.R [hunk 4]
```diff
@@ -121,3 +121,45 @@ sf_to_zip <- function(zip_filename, sf_object, layer_name){
...
+  second_res <- data.frame(date = seq(min(obs_dat$date[obs_dat$site_id %in% unique(obs_dat$site_id)[2]]), 
+                                      max(obs_dat$date[obs_dat$site_id %in% unique(obs_dat$site_id)[2]]), by = 1)) %>%
+    full_join(obs_dat[obs_dat$site_id %in% unique(obs_dat$site_id)[2],], by = "date") %>%
+    mutate(site_id = unique(obs_dat$site_id)[2])
+  
+  out_dat <- bind_rows(first_res, second_res) %>%
+    arrange(site_id, date) %>%
+    group_by(site_id) %>%
+    mutate(surface_elevation_m = na.approx(surface_elevation_m)) %>%
+    mutate(data_type = ifelse(is.na(data_type), 'daily interpolated', data_type))
```
Comment (hcorson-dosch-usgs) [inline src/file_functions.R:161]: I like this addition of a `data_type` column in addition to the `source` column

**Review by hcorson-dosch-usgs — APPROVED** [review]
Looks good, Sam! Once this is merged I'll rebuild everything and push to SB, then updated the fetching in `res-temperature-process-models`

File: src/file_functions.R (hunk 3 above)
Comment (aappling-usgs) [reply src/file_functions.R:138]: Ooh, yeah, that's news to me too! Nice!

File: 2_observations.yml (hunk 1 above)
Comment (limnoliver) [reply 2_observations.yml]: ahh, just oversight, I think. I was using the `fetch_filter_tibble` as a template, and didn't make the connection that I don't need both!

File: src/file_functions.R (hunk 2 above)
Comment (limnoliver) [reply src/file_functions.R:137]: Yeah - this is one place where there were lots of ways to do what I wanted, and having numbers from the `.id` argument was efficient, but not the easiest to understand when reading the code.

File: src/file_functions.R (hunk 4 above)
Comment (limnoliver) [reply src/file_functions.R:161]: Hoping this is helpful to the modelers if anything seems wonky down the road!
